### DISCUSSION TITLE
Tested Terminals

### DISCUSSION THREAD

**Discussion by o-sdn-o** [directvt/vtm#72] [General]
GUI mode is available and automatically enabled on Windows and Unix-like platforms with X11/XWayland. On macOS, a terminal emulator is currently required. To force `vtm` to launch in TUI mode, use the command-line argument `vtm --tui`.

> Note: vtm itself has a built-in terminal emulator that can be launched via `vtm -r term [your_shell]`.

### Tested Terminal Emulators

 - [Termux](https://github.com/termux/termux-app)
 - [tmux](https://github.com/tmux/tmux)
    - Release of the second mouse button is not reported when two mouse buttons are pressed simultaneously
 - [Ultimate++ Terminal widget](https://github.com/ismail-yilmaz/upp-components/tree/master/CtrlLib/Terminal)
 - [Gnome Terminal](https://wiki.gnome.org/Apps/Terminal)
 - [Linux VGA Console (Linux in-kernel console)](https://en.wikipedia.org/wiki/Linux_console)
    - 16-color mode (8 background, 16 foreground)
    - In order to use a mouse or touchpad in Linux VGA Console, you must grant the user access to mouse/pointing devices. By default, only privileged users and users of the `input` group have access. To grant all users permanent access to all pointing devices, use the command:
      - ```
        sudo vtm --mouse
        ```
      To reset permanent access, use the command:
      - ```
        sudo vtm --mouse 0
        ```
 - [KMSCON](https://github.com/Aetf/kmscon) v9 and higher.
   - In order to use a mouse or touchpad you need to do the same steps as for Linux VGA Console.
   - KMSCON doesn't report the `Ctrl + PgDn/PgUp` key press events for some reasons.
 - [yaft](https://github.com/uobikiemukot/yaft)
   - In order to use a mouse or touchpad you need to do the same steps as for Linux VGA Console.
   - yaft doesn't report the `Ctrl + PgDn/PgUp` key press events for some reasons.
 - [st, stterm](https://st.suckless.org/)
 - [mintty](https://github.com/mintty/mintty)
 - [Windows Terminal](https://github.com/microsoft/terminal)
 - [Windows Command Prompt](https://github.com/microsoft/terminal)
    - `Windows 8.1`/`Windows Server 2012` or later
 - [kitty](https://github.com/kovidgoyal/kitty)
 - [ghostty](https://ghostty.org/)
 - [xterm](https://invisible-island.net/xterm/)
 - [Alacritty](https://github.com/alacritty/alacritty)
 - [xfce4-terminal](https://docs.xfce.org/apps/terminal/start)
 - [Konsole](https://konsole.kde.org/)
 - [wezterm](https://wezfurlong.org/wezterm/index.html)
 - [iTerm2](https://iterm2.com/)
   - It is necessary to assign shortcuts `Ctrl + PgDn/PgUp`  and unbind the `Right button single click`  <details><summary>Screenshots...</summary><p>
     - ![iterm2_pgup](https://user-images.githubusercontent.com/11535558/126869068-db4baf34-ad75-40b8-8b1f-9e9802b254b5.png)
     - ![iterm_pgdown](https://user-images.githubusercontent.com/11535558/126869070-8defc768-b84b-4f29-9c7b-db65bc5f7bd7.png)
     - ![iTerm2_rightclick](https://user-images.githubusercontent.com/11535558/126869129-531db036-64ad-40ab-807c-8ecb48c8fd62.png)
</p></details>

 - [Apple Terminal, Terminal.app](https://en.wikipedia.org/wiki/Terminal_(macOS))
   - [Release of the second mouse button is not reported when two mouse buttons are pressed simultaneously](https://feedbackassistant.apple.com/feedback/9393516)
   - 256-color mode (16 ANSI + 6×6×6 cube + 24 grayscale)
   - It is necessary to assign shortcuts `Ctrl + PgDn/PgUp`<details><summary>Screenshots...</summary><p>
     - ![Terminal app_pgup](https://user-images.githubusercontent.com/11535558/126869412-581776e8-8056-44da-8d3d-4c74b6921485.png)
     - ![Terminal app_pgdn](https://user-images.githubusercontent.com/11535558/126869415-c11615f2-bd85-4a2a-bc99-55a6ad058264.png)
</p></details>

 - [Tilix](https://gnunn1.github.io/tilix-web/)
   - Shortcuts `Ctrl + PgDn/PgUp` are reserved for switching between sessions. Switching between windows doesn't work without reassigning these shortcuts.
 - [Konsole](https://konsole.kde.org/)
   - DECSET 1003 mouse tracking mode is supported since [07/06/2021](https://bugs.kde.org/show_bug.cgi?id=391967)
 - [PuTTY](https://www.putty.org/)
    - DECSET 1002 and 1006 are supported by PuTTY, which is enough to move windows, but not enough to highlight active UI elements
 - [qterminal](https://github.com/lxqt/qterminal)
    - Swapped right and middle buttons

**Comment by o-sdn-o** [comment]
The core problem

Without support for character fragmentation, wide characters cannot be displayed in half (e.g. halves with different background colors)
![halves_00](https://user-images.githubusercontent.com/11535558/92585399-560ac880-f2ae-11ea-9263-ce16b9f40619.gif)

A possible solution is the use of some kind of escape sequence to display fragments of characters  
https://gitlab.freedesktop.org/terminal-wg/specifications/-/issues/23

- 5/13/2024: It is going to be fixed soon. GUI mode: https://youtu.be/ul2Y8S5WT1Q.
- 7/04/2024: That's not a problem today since [v0.9.84](https://github.com/directvt/vtm/releases/tag/v0.9.84).

**Comment by USLTD** [comment]
**tmux** isn't terminal, is it? 🧐
Also please add **Termux** to tested terminals. Only problem is dragging on mobile but I don't know about with physical mouse

**Reply by o-sdn-o** [reply]
tmux is a full fledged terminal emulator.

[Wikipedia](https://en.wikipedia.org/wiki/Terminal_emulator): Terminal emulator, is a computer program that emulates a video terminal within some other display architecture 🙂

**Reply by USLTD** [reply]
I don't agree with that.
**tmux** is not *Terminal Emulator* but a *Terminal Multiplexer*. It's a big difference and *Terminal Emulator* ≠ *Terminal Multiplexer*

**Reply by o-sdn-o** [reply]
In my mind, Not every terminal emulator is a multiplexer, but every multiplexer is a terminal emulator.

**Reply by o-sdn-o** [reply]
Any multiplexer can be launched both inside the common terminal emulator and recursively inside itself.

**Reply by USLTD** [reply]
I regard **Terminal Multiplexer** as **Pseudo-terminal Manager** rather than **Emulator**. With what you said **Terminal Multiplexers** look like **Emulators** but I still think that it's kinda wrong 😓
IDK.

**Reply by o-sdn-o** [reply]
Terminal emulators and multiplexers have identical internals.  Their difference is only in the form in which the output occurs.  A terminal emulator can have many panes and tabs and the resulting output is produced in the GUI.  A terminal multiplexer has everything the same, but the resulting output produced in the TUI.

**Reply by USLTD** [reply]
Ok. Also why don't add versions of Terminals you tested? For example my **Termux**'s version is **0.118.0**. I think it will be helpful

**Reply by o-sdn-o** [reply]
I decided not to specify terminal versions because it is unlikely that any terminal will stop supporting truecolor or UTF-8 (minimum requirements) if it currently supports it.  If vtm does not work correctly in some of the terminals listed above, then the only solution is to update this terminal to the latest version.

**Reply by o-sdn-o** [reply]
vtm actually has very modest requirements for terminals, and if it works fine in something, then it will always work fine in it

**Comment by megatog615** [comment]
Can't seem to get mouse input working on Debian Stable in the Linux Console. I am trying to set this up on an old i686 Pentium III system and have already compiled and confirmed that it runs on the system. It doesn't work on other Debian boxes that I have running, even with chmodding /dev/input/mice to 666. I don't use gpm.

Everything seems to work just fine in the terminal of my choice, though. Would be great to get this working correctly on the PIII system as it would be its primary desktop mode.

**Reply by o-sdn-o** [reply]
Can you please tell me what model of mouse you are using?

vtm when running in the Linux console only supports mice compatible with the ImPS/2 protocol.  I can add support for other protocols if that's the case.

**Reply by o-sdn-o** [reply]
`ImPS/2` = `Microsoft Intellimouse PS/2 Extension`, https://isdaman.com/alsos/hardware/mouse/ps2interface.htm

**Reply by megatog615** [reply]
I am not sure how to check. The system is running Debian Stable i386. It is a Panasonic Toughbook CF-M34 and it has a touchpad that I am assuming is wired using PS/2. The following relevant device nodes are available:
```
/dev/input/mice
/dev/input/mouse0
/dev/psaux
```

Plugging in a USB mouse strangely doesn't work either. Running vtm as root also doesn't have a working mouse.

**Reply by o-sdn-o** [reply]
I suspect that this laptop is equipped with Synaptic Touchpad, I will try to implement support for it.

**Reply by megatog615** [reply]
From running Xorg in the past I can confirm that it does use the synaptics driver. Modern Xorg is way too slow to run on this machine so I don't use it anymore.

If Xorg can detect a PS/2 device surely there must be a utility that can detect and list off these devices? If only it was as easy as 'lsusb' or 'lspci'.
EDIT: Oh my, there is(**lshw**).
Output of lshw suggests it's PNP device 0f13 which is an ALPS touchpad.

**Reply by o-sdn-o** [reply]
`Synaptic` has a compatibility mode, maybe `AlpsPS/2 ALPS` has the same functionality, I google their capabilities.

**Reply by o-sdn-o** [reply]
I've added mouse support without a scroll wheel.  Perhaps this will be enough.  I cannot test this for now.  Could you check if this option works (dev/sdn branch).

**Reply by megatog615** [reply]
It works! I'll return post again if I find any issues with it. This pretty much enables VTM on retro 90's (PS/2-compatible) PCs!

**Reply by o-sdn-o** [reply]
Thank you!  This will really work in full force when the size of allocated memory is reduced.  So far, memory consumption is not optimized, and it can be reduced by 30 times.

**Comment by ldemailly** [comment]
ghostty?

also a bit curious about the overall... goal

**Reply by o-sdn-o** [reply]
The main goal is to create a new class of applications -- something between a GUI and a TUI-capable of running both in their own GUI windows and in any text console (vtm itself already supports this out of the box on Windows). Applications whose framework is built on DynamicXML+Lua will be similar to WPF (with XAML+CLR) or WebApp (with HTML+JS). Here, vtm itself is the environment and testing ground for the required technologies --DynamicXML, Lua-scripting, sophisticated input (keyboard, mouse, touchpad, and so on -- I rewrote the entire libinput library in C++ for this -- see lixx.hpp).

**Reply by o-sdn-o** [reply]
You can feel the difference on Windows:
- `vtm --tui --run test`
- `vtm --gui --run test`

**Reply by o-sdn-o** [reply]
vtm has the ability to fully deserialize/serialize its state through arbitrary channels (like socat over ssh reverse tunnel) so you can also do things like this:

> ### Remote Access using `socat`/`winsocat` over SSH Reverse Tunnel
> 
> Notes:
> - The remote and local sides may differ in platform.
> - Using the **localhost** name may cause connection issues with the IP address family (IPv4: `127.0.0.1` vs. IPv6: `[::1]`) between remote systems. Use an explicit IP address to avoid this.
> 
> #### Unix
> 
> - Remote side
>   ```bash
>   #   Run reverse tunnel for 1122tcp in background
>   #                         ┌─ relay all traffic passed through 127.0.0.1:1122tcp to the remote host
>   #                ─────────┴────
>     ssh -N -R 1122:127.0.0.1:1122 user@1.2.3.4 &
>   #     ─┬ ─┬─────                ─┬────────── ┬
>   #      │  └─ listen on remote    │           └─ run in background
>   #      │                         └─ remote host
>   #      └─ do nothing on remote just listen 1122tcp
>   #
>   # Save the background process'' PID
>   ssh_tunnel_pid=$!
>   #
>   # Run vtm and forward its stdio (in dtvt format) streams to 127.0.0.1:1122
>   #            ┌──────< relay dtvt-stream <─────┐      ┌─ kill the ssh reverse tunnel after vtm exits
>   #       ─────┴────────────────────────       ─┴─   ──┴─────────────────
>     socat tcp-listen:1122,bind=127.0.0.1 exec:"vtm"; kill $ssh_tunnel_pid
>   #       ─────┬────────────────────────
>   #            └─ wait forwarded connections on 127.0.0.1:1122tcp
>   # or reversed
>   # socat exec:"vtm" tcp-listen:1122,bind=127.0.0.1; kill $ssh_tunnel_pid
>   ```
> - Local side with ip=1.2.3.4 (connect to the remote vtm desktop via 127.0.0.1:1122tcp)
>   ```bash
>   #    ┌──────────< binary DirectVT <─────────────┐
>   # ───┴───────                          ─────────┴─────
>     vtm -r dtvt socat tcp:127.0.0.1:1122 stdin\!\!stdout
>   #                   ────┬───────────── ─────────┬─────
>   #                       └─> relay dtvt-stream >─┘
>   ```
> 
> #### Windows
> 
> - Install `winsocat`:
>   ```
>   winget install winsocat
>   ```
> - Remote side
>   ```pwsh
>   # Run reverse tunnel for 1122tcp in background.
>   $process = Start-Process "ssh" "-N -R 1122:127.0.0.1:1122 user@1.2.3.4" -PassThru -WindowStyle Minimized
>   # Run vtm and forward its stdio streams to 127.0.0.1:1122. Kill the ssh reverse tunnel after vtm exits.
>   #               ┌──────< relay dtvt-stream <─────┐      ┌─ kill the ssh reverse tunnel after vtm exits
>   #          ─────┴────────────────────────       ─┴─   ──┴─────────────────────────
>     winsocat tcp-listen:1122,bind=127.0.0.1 exec:"vtm"; Stop-Process -Id $process.Id
>   #          ─────┬────────────────────────
>   #               └─ wait forwarded connections on 127.0.0.1:1122tcp
>   # or
>   # winsocat exec:"vtm" tcp-listen:1122,bind=127.0.0.1; Stop-Process -Id $process.Id
>   ```
> - Local side with ip=1.2.3.4 (connect to the remote vtm desktop via 127.0.0.1:1122tcp)
>   ```pwsh
>   #    ┌───────< binary DirectVT <─────────────┐
>   # ───┴───────                             ───┴─
>     vtm -r dtvt winsocat tcp:127.0.0.1:1122 stdio
>   #                      ┬───────────────── ───┬─
>   #                      └> relay dtvt-stream >┘
>   # Kill the abandoned winsocat process (it keeps running for some reason)
>   Get-Process -Name "winsocat" | Stop-Process -Force
>   ```

**Reply by o-sdn-o** [reply]
Another huge argument against ghostty (library) is the requirement for [vt2d](https://github.com/directvt/vtm/blob/master/doc/character_geometry.md) support.

**Reply by ldemailly** [reply]
I had 2 different questions in my comment above and maybe it got confusing:

question 1: is ghostty one of the supported terminal (like you list kitty, alacritty etc) I was suggesting to update this list to add ghostty (as I assume it'd work about same as kitty)

question 2 was my confusion about what is "vtm" which you kindly answered despite being off topic (my bad)

**Reply by o-sdn-o** [reply]
I may have confused the library (libghosty) and the terminal. For some reason, I thought the question was why not use a library for compatibility with all terminals. This is most likely a question about vtm compatibility with Ghostty. It's probably not listed here simply because vtm works fine in Ghostty. Excuse me, it's 4:15 AM.

**Reply by o-sdn-o** [reply]
we wrote at the same time 😄
